crates/client/eth/src/l1_gas_price.rs

```diff
@@ -0,0 +1,235 @@
+) -> anyhow::Result<()> {
+    let poll_time = eth_client.gas_price_poll_ms.unwrap_or(DEFAULT_GAS_PRICE_POLL_MS);
+    update_last_update_timestamp().await;
+    loop {
+        match update_gas_price(eth_client, Arc::clone(&l1_data_provider)).await {
+            Ok(_) => log::trace!("Updated gas prices"),
+            Err(e) => log::error!("Failed to update gas prices: {:?}", e),
+        }
+
+        let last_update_timestamp = get_last_update_timestamp().await;
+        let current_timestamp = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("Failed to get current timestamp")
+            .as_millis();
+
+        if current_timestamp - last_update_timestamp > 10 * poll_time as u128 {
+            panic!(
+                "Gas prices have not been updated for {} ms. Last update was at {}",
+                current_timestamp - last_update_timestamp,
+                last_update_timestamp
+            );
+        }
+
+        if !infinite_loop {
+            return Ok(());
+        }
+
+        sleep(Duration::from_millis(poll_time)).await;
+    }
+}
+
+async fn update_gas_price(
+    eth_client: &EthereumClient,
+    l1_data_provider: Arc<dyn L1DataProvider>,
+) -> anyhow::Result<()> {
+    let fee_history = eth_client.provider.get_fee_history(300, BlockNumberOrTag::Latest, &[]).await?;
+
+    // The RPC responds with 301 elements for some reason. It's also just safer to manually
+    // take the last 300. We choose 300 to get average gas caprice for last one hour (300 * 12 sec block
+    // time).
+    let (_, blob_fee_history_one_hour) =
+        fee_history.base_fee_per_blob_gas.split_at(fee_history.base_fee_per_blob_gas.len().max(300) - 300);
+
+    let avg_blob_base_fee = blob_fee_history_one_hour.iter().sum::<u128>() / blob_fee_history_one_hour.len() as u128;
+
+    let eth_gas_price = fee_history.base_fee_per_blob_gas.last().context("Setting l1 last confirmed block number")?;
+
+    l1_data_provider.update_eth_l1_gas_price(*eth_gas_price).await;
+    l1_data_provider.update_eth_l1_data_gas_price(avg_blob_base_fee).await;
+
+    update_last_update_timestamp().await;
+
+    // Update block number separately to avoid holding the lock for too long
+    update_l1_block_metrics(eth_client, l1_data_provider).await?;
+
+    Ok(())
+}
+
+async fn update_l1_block_metrics(
+    eth_client: &EthereumClient,
+    l1_data_provider: Arc<dyn L1DataProvider>,
+) -> anyhow::Result<()> {
+    // Get the latest block number
+    let latest_block_number = eth_client.get_latest_block_number().await?;
+
+    // Get the current gas price
+    let current_gas_price = l1_data_provider.get_gas_prices().await;
+    let eth_gas_price = current_gas_price.eth_l1_gas_price;
+
+    // Update the metrics
+    eth_client.l1_block_metrics.l1_block_number.set(latest_block_number as f64);
+    eth_client.l1_block_metrics.l1_gas_price_wei.set(eth_gas_price as f64);
+
+    // We're ignoring l1_gas_price_strk
```

Comment:
can we import the strk code as well? i remember pragma did a PR. or do we want to do it in another issue

Reply:
we would need an oracle for that right? I will check for the PR, it can be another PR with the scope of adding oracle